src/httpClient.js

```diff
@@ -1,98 +1,97 @@
 const httpClient = {
-  values: {},
-  defaultValuesFunctions: {},
-
-  value: function(name) {
-    if (this.values[name]) {
-      return this.values[name]
-    } else if (this.defaultValuesFunctions[name]) {
-      return this.defaultValuesFunctions[name]()
-    } else {
-      throw "value " + name + " not found"
+    values: {},
+    defaultValuesFunctions: {},
+
+    value: function(name) {
+        if (this.values[name]) {
+            return this.values[name]
+        } else if (this.defaultValuesFunctions[name]) {
+            return this.defaultValuesFunctions[name]()
+        } else {
+            throw "value " + name + " not found"
+        }
+    },
+
+    defineValue: function(key, defaultFunction) {
+        defaultFunction = defaultFunction || function () { return null }
+
+        this[key] = function (value) { return this.newWithValue(key, value) }
+        this.defaultValuesFunctions[key] = defaultFunction
+    },
+
+    defineMergedValue: function(key, defaultFunction) {
+        defaultFunction = defaultFunction || function () { return {} }
+        this.defineValue(key, defaultFunction)
+
+        this[key] = function (subValues) {
+            return this.newWithMergedValue(key, subValues)
+        }
+    },
+
+    newWithValue: function(key, value) {
+        const newValues = Object.assign({}, this.values)
+        newValues[key] = value
+
+        return Object.assign({}, this, {values: newValues})
+    },
+
+    newWithMergedValue: function(key, object) {
+        const oldObject = this.value(key)
+        const newObject = Object.assign({}, oldObject, object)
+        return this.newWithValue(key, newObject)
+    },
+
+    findCallbackByStatus: function (resp) {
+        return this.value('onStatusCallbacks')[resp.status]
+    },
+
+    findCallbackBySuccess: function (resp) {
+        return resp.ok ? this.value('onSuccess') : this.value('onErrorResponse')
+    },
+
+    request: function (url, opts) {
+        return this
+            .url(url)
+            .requestOptions(opts)
+    },
+
+    runRequest: function (url, opts) {
+        return this.request(url, opts).run()
+    },
+
+    run: function () {
+        return this.value('onBeforeRun')(this).doRun()
+    },
+    doRun: function () {
```

Comment:
changes here

Reply:
Mario, please don't change the indentation of the whole file. It's very difficult to know what changed. And if we finally agree that four spaces are better, at least commit that change separately so changes can be compared individually.